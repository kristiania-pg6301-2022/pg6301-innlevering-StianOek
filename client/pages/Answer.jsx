import React, { useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate } from "react-router-dom";

const Answer = ({ data }) => {
  const navigate = useNavigate();

  if (!data) return <h1>LOADING....</h1>;

  let correct = data.correct;
  let answers = data.answers;

  return (
    <div>
      <Routes>
        <Route path={"correct"} element={<h1>Correct!</h1>} />
        <Route path={"wrong"} element={<h1>Wrong!</h1>} />
      </Routes>
      <div>
        <h2 data-testid={"status"}>
          Your score is {correct} / {answers}
        </h2>
      </div>
      <div>
        <Link to="/">
          <button>Return to home</button>
        </Link>
        <Link to="/question">
          <button>New quiz</button>
        </Link>
      </div>
    </div>
  );
};

export default Answer;
