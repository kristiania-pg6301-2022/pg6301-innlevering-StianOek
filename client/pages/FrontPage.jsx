import React, { useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate } from "react-router-dom";

const FrontPage = () => {
  return (
    <div>
      <div>
        <h1>Do you master javascript?</h1>
        <h2>Take a test</h2>
      </div>
      <div>
        <Link to="/question">
          <button>New quiz</button>
        </Link>
        <Link to="/answer/*">
          <button>Show anwers</button>
        </Link>
      </div>
    </div>
  );
};
export default FrontPage;
