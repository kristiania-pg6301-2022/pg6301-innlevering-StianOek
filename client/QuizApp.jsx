import React from "react";
import { Link, Route, Routes, useNavigate } from "react-router-dom";
import { fetchJSON, postJSON, useLoading } from "./useFetch";

/* Imported pages */
import FrontPage from "./pages/FrontPage";
import Answer from "./pages/Answer";

export const Question = ({ reload, fetchQuestion }) => {
  const navigate = useNavigate();
  const { data: question, error, loading } = useLoading(fetchQuestion);

  if (error) {
    return (
      <p>
        An error has occured:{" "}
        <span style={{ color: "red" }}>{error.toString()}</span>
      </p>
    );
  }

  if (!question) {
    return <h1>Loading...</h1>;
  }

  const handleRightAnswer = async (answers) => {
    const id = question.id;
    const res = await fetch("/api/answer", {
      method: "post",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ id, answers }),
    });
    const result = await res.json();

    reload();

    if (result.result === "incorrect") {
      navigate("/answer/wrong");
    } else if (result.result === "correct") {
      navigate("/answer/correct");
    }
  };

  return (
    <div>
      <div>
        <h1>{question.question}</h1>
        <h2>Category: {question.category}</h2>
      </div>
      {Object.keys(question.answers)
        .filter((answer) => question.answers[answer])
        .map((value) => {
          return (
            <div key={value}>
              <button onClick={() => handleRightAnswer(value)}>
                {question.answers[value]}
              </button>
            </div>
          );
        })}
    </div>
  );
};

const QuizApp = () => {
  const { reload, data, error, loading } = useLoading(
    async () => await fetchJSON("/api/score")
  );
  const fetchQuestion = async () => await fetchJSON("/api/random");

  return (
    <div>
      <Routes>
        <Route path={"/"} element={<FrontPage />} />
        <Route
          path={"/question"}
          element={<Question reload={reload} fetchQuestion={fetchQuestion} />}
        />
        <Route path={"/answer/*"} element={<Answer data={data} />} />
      </Routes>
    </div>
  );
};

export default QuizApp;
