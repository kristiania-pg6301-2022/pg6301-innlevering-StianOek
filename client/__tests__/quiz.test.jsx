import React from "react";
import ReactDOM from "react-dom";
import FrontPage from "../pages/FrontPage";
import { MemoryRouter } from "react-router-dom";
import Answer from "../pages/Answer";
import { act, Simulate } from "react-dom/test-utils";

import { Question } from "../QuizApp.jsx";

describe("QuizApp", () => {
  it("should show frontpage", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <FrontPage />
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should check for answers", () => {
    const element = document.createElement("div");

    ReactDOM.render(
      <MemoryRouter>
        <Answer data={{ correct: 2, answers: 5 }} />
      </MemoryRouter>,
      element
    );
    expect(element.querySelector("[data-testid=status]").textContent).toEqual(
      "Your score is 2 / 5"
    );
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should fetch random questions", async () => {
    const container = document.createElement("div");
    document.body.appendChild(container);

    const reload = jest.fn();

    const fetchQuestion = async () =>
      await {
        id: 974,
        category: "Code",
        question:
          "What is the correct JavaScript syntax to change the content of the HTML element below?",
        answers: {
          answer_a: '#demo.innerHTML = "Hello World!";',
          answer_b:
            'document.getElementById("demo").innerHTML = "Hello World!";',
          answer_c: 'document.getElement("p").innerHTML = "Hello World!";',
          answer_d:
            'document.getElementByName("p").innerHTML = "Hello World!";',
        },
      };

    await act(async () => {
      ReactDOM.render(
        <MemoryRouter initialEntries={["/question"]}>
          <Question fetchQuestion={fetchQuestion} />
        </MemoryRouter>,
        container
      );
      // Simulate.click(container.querySelector("[data-testid=answer_a] button"));
    });

    expect(container.innerHTML).toMatchSnapshot();
    //expect(reload).toBeCalled();
    expect(container.querySelector("h1").textContent).toEqual(
      "What is the correct JavaScript syntax to change the content of the HTML element below?"
    );
  });
});
