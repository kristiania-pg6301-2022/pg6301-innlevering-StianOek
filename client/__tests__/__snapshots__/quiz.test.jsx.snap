// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizApp should check for answers 1`] = `"<div><div><h2 data-testid=\\"status\\">Your score is 2 / 5</h2></div><div><a href=\\"/\\"><button>Return to home</button></a><a href=\\"/question\\"><button>New quiz</button></a></div></div>"`;

exports[`QuizApp should fetch random questions 1`] = `"<div><div><h1>What is the correct JavaScript syntax to change the content of the HTML element below?</h1><h2>Category: Code</h2></div><div data-testid=\\"answer_a\\"><button>#demo.innerHTML = \\"Hello World!\\";</button></div><div data-testid=\\"answer_b\\"><button>document.getElementById(\\"demo\\").innerHTML = \\"Hello World!\\";</button></div><div data-testid=\\"answer_c\\"><button>document.getElement(\\"p\\").innerHTML = \\"Hello World!\\";</button></div><div data-testid=\\"answer_d\\"><button>document.getElementByName(\\"p\\").innerHTML = \\"Hello World!\\";</button></div></div>"`;

exports[`QuizApp should show frontpage 1`] = `"<div><div><h1>Do you master javascript?</h1><h2>Take a test</h2></div><div><a href=\\"/question\\"><button>New quiz</button></a><a href=\\"/answer/*\\"><button>Show anwers</button></a></div></div>"`;
